Introduce DimAnalysis class for unknown dimension analysis and apply it to NNPA

tungld opened the pull request:
This patch introduces a **DimAnalysis** class to analyze unknown dimensions in ONNX IR.
In particular, it classifies unknown dimensions into groups/sets. Dimensions in a group are supposed to be equal at runtime time. Each group has an unique ID.

ShapeHelper is used to explore the equality relationship between input and output unknown dimensions.

A pass `--onnx-dim-analysis` is introduced for the testing purpose where it emits `onnx.DimGroup` to the IR to show group IDs.

This patch only supports a limited number of operations (e.g. operations in ResNet50 and BertSquad). More supported operations will be added in future PRs.

As an example, this patch shows how to apply DimAnalysis to ONNXToZHigh pass where we identify if there is broadcasting in binary operations such as Add/Div/Mul/Sub or not. Applying to ResNet50 whose input is `?x3x224x224`, All 16 ONNXAdd ops are identified as non-broadcasting ops and are lowered to zHigh for NNPA.

The number of operations in ResNet50 when **not** using DimAnalysis:
<img width="282" alt="Screen Shot 2022-11-01 at 12 28 47" src="https://user-images.githubusercontent.com/1230113/199152352-2d3c9da3-25ca-49ae-a357-ff6756ba170c.png">

The number of operations in ResNet50 when using DimAnalysis:
<img width="281" alt="Screen Shot 2022-11-01 at 21 02 22" src="https://user-images.githubusercontent.com/1230113/199228326-16093f39-119e-4b08-9ce8-b8038f5b68f1.png">

AlexandreEichenberger commented on src/Dialect/ONNX/ShapeInference/Pad.cpp:
```diff
@@ -60,7 +60,12 @@ LogicalResult ONNXPadOpShapeHelper::computeShape(
     // Calculation for output size.
-    IndexExpr dimOutputFinal = padBegin + dimInput + padEnd;
+    IndexExpr dimOutputFinal;
+    if (padBegin.isLiteral() && padEnd.isLiteral() &&
```
Is the problem that if they are both zero, and the input dim is question mark, that is the only way to preserve the same question mark? 
I think we should change the propagation rules that `question mark + 0` should preserve the question mark in the index expressions. Then you can do that systematically, `? * 1`, `? / 1`, `select(true, ?, x)`...
Then we don't have this exception here, and it would apply to many many more cases.

tungld replied:
Yes, I am aware of this and it's interesting enough to be addressed with a new PR. So, I just do a quick fix for PadOp that is used in ResNet50.

AlexandreEichenberger replied:
Sure, so add a TODO to remove this handling once the proper arithmetic is in.

AlexandreEichenberger reviewed (commented):
First rounds of comments, will read the `ONNXDimAnalysis` files next

AlexandreEichenberger commented on src/Transform/ONNX/ONNXDimAnalysis.hpp:
```diff
@@ -0,0 +1,72 @@
+
+class DimAnalysis {
+public:
+  using DimT = std::pair<mlir::Value, uint64_t>;
```
Can you indicate for each type what they are used for. That would help the comprehension.

AlexandreEichenberger replied:
tx

AlexandreEichenberger commented on src/Transform/ONNX/ONNXDimAnalysis.hpp:
```diff
@@ -0,0 +1,72 @@
+  /// Returns the grouping result of unknown dimensions.
+  DimSetMapT getGroupingResult() const { return dimSetMap; }
+
+  /// Test if two unknown dimensions are the same or not.
```
what are `lhs` and `rhs`? Are they value of a shaped type kind? So `tensorOrMemref`? If so, a more explicit name would facilitate the usage of the interface.

AlexandreEichenberger commented on src/Transform/ONNX/ONNXDimAnalysis.cpp:
```diff
@@ -0,0 +1,588 @@
+// Helper functions for dimension analysis.
+//===----------------------------------------------------------------------===//
+
+/// Check if two sets are interleaved or not.
```
Interleave, do you mean do the the two sets intersect? A set based name would be clearer, maybe `nonEmptyIntersection`

AlexandreEichenberger commented on src/Transform/ONNX/ONNXDimAnalysis.cpp:
```diff
@@ -0,0 +1,588 @@
+    mlir::ValueRange operands, onnx_mlir::DimAnalysis::DimSetT &sameDims) {
+  int64_t outputQMValue = qmOuputIE.getQuestionmark();
+  // Cannot process if the question mark value is -1 that is a common value for
+  // every question mark.
```
I would abstract here a bit the interface.
```
bool qmOutputIE.specificQuestionmark() 
bool qmOutputIE.sameQuestionmark(iIR)
```
whereas the first one assert if IE is not a question mark, and return true if `!= -1`,
and whereas the second one assert if either is not a question mark, and return true if their values are not `-1` and are the same (negative) value.

AlexandreEichenberger replied:
I believe this comment was not resolved, at least as I see it in the code below

tungld replied:
Done. I created the two functions but forgot using them. Thanks!

AlexandreEichenberger commented on src/Transform/ONNX/ONNXDimAnalysis.cpp:
```diff
@@ -0,0 +1,588 @@
+      }
+    }
+  }
+  return;
```
no need for return

tungld replied:
Done. Thanks!

AlexandreEichenberger commented on src/Transform/ONNX/ONNXDimAnalysis.cpp:
```diff
@@ -0,0 +1,588 @@
+/// same unknown dimensions in the inputs.
+void findAndAddSameDim(const onnx_mlir::QuestionmarkIndexExpr &qmOuputIE,
+    mlir::ValueRange operands, onnx_mlir::DimAnalysis::DimSetT &sameDims) {
+  int64_t outputQMValue = qmOuputIE.getQuestionmark();
```
Ideally this call would not be needed, because we only query the index expression directly for the info we need. See below.

tungld replied:
Could you elaborate a bit? I don't see "below" things, maybe you forgot pushing them ...

AlexandreEichenberger replied:
To the extend possible, we should not need the `getQuestionmark()` call.
We should have
```
bool qmOutputIE.specificQuestionmark() 
bool qmOutputIE.sameQuestionmark(iIR)
```
where we ask index expression if an index expression question mark is specific (<-1) or not (==-1), and if they are the same (same specific number).

If we were to change the representation, it would not impact the code here.

tungld replied:
Thanks for th explaination! I addressed this just now.

tungld commented on src/Accelerators/NNPA/Conversion/ONNXToZHigh/ONNXLegalityCheck.cpp:
```diff
@@ -37,10 +38,11 @@ bool isValidElementType(Value val) {
+/// dimension analysis to try inferring the equality at compile.
 /// Also, check the ranks of two tensors, they must be in range of (0, 4].
-bool haveSameStaticShape(Value value1, Value value2) {
+bool haveSameShape(
```
I put a TODO here to find a new home for it. I tried to move it the ONNX Ops helper but it caused cyclic dependency: ONNXOp -> DimAnalysis -> ONNXOps -> ...

AlexandreEichenberger replied:
The obvious spot would be inside DimAnalysis.

tungld replied:
I moved it into DimAnalysis. Thanks!

tungld commented on src/Dialect/ONNX/AdditionalONNXOps.td:
```diff
@@ -158,14 +158,39 @@ def ONNXDimOp : ONNX_Op<"Dim",
+    Value -1 for `group_id` means the dimension does not belong to any group.
+
+    This operation is not part of the standard and was added to assist onnx-mlir.
+  }];
```
Added.

AlexandreEichenberger replied:
thanks.

tungld commented:
@AlexandreEichenberger thanks for your comments. I addressed most of them except the last one I am a bit confused. I also added a document on how to use DimAnalysis and more comments to code as @chentong319 suggested. It's ready for another round of review.

AlexandreEichenberger commented on docs/UnknownDimensionAnalysis.md:
```diff
@@ -0,0 +1,30 @@
+It is often the case where we want to know two unknown dimensions are equal or not at compile time. This helps with decision on how to lowering an ONNX operator. For example, given an ONNXAddOp as follows:
```
Awesome, thanks for adding this

AlexandreEichenberger reviewed (approved):
LGTM, the text of a comment in IndexExpr needs to be corrected, otherwise all good

tungld commented on src/Dialect/Mlir/IndexExpr.hpp:
```diff
@@ -650,6 +650,17 @@ class QuestionmarkIndexExpr : public IndexExpr {
+  // other question marks.
+  bool specificQuestionmark() const;
+  // Check if two question marks are the same or not. Two question marks are the
+  // same if they are not specific and have the same value.
```
Good catch! Fixed it, thanks!
